### ISSUE TITLE
Initial experiments on staged multilingual training

### ISSUE THREAD

**Issue by spyysalo** [OpenEuroLLM/Taskboard#23]
First experiment studying the impact of training multilingually from step 1 vs. later in training

**Comment by alex-shvets** [comment]
We are planning to test several approaches. I put here a draft of our ideas that we'll refine soon.

**EXPERIMENT SET 1 (Staged training)**
Train 3B models from scratch to test different data mix strategies (with sub-strategies), so that we can compare different performances.

- **100% English - 0% multilingual then swap**
    - First stage training with fully English data → a translation-intensive and also a bit of multilingual and English training → multilingual training
    - First stage training with fully English data → multilingual training
    - **This way we compare the importance of the translation-intensive step**
- **80%/20% trainings**
    - 80% English and 20% multilingual → swap, 20% English and 80% multilingual
    - Same as above but add another intermediate step(s), to do it more gradually
    - **This way we compare the importance of smooth or jump adaptation**

**EXPERIMENT SET 2 (multilingual adaptation)**
This idea is not so advanced yet. But we would be interested in trying something similar to the idea presented in https://arxiv.org/abs/2410.01335

Some ideas that could be implemented:
- **all domains:** Intensive English train full LLM → translation intensive train only lower and upper layers → multilingual training only upper and lower layers
- **math specific:** Domain specific training (math) of full LLM → translation training only lower and upper → one language training lower and upper → check specific field in new language
- **Progressive layer unfreezing schedule:** Train full LLM using English →freeze core 50% and train top and bottom 25% → freeze core 30% and train 35% top and bottom → unfreeze all
- **Reversed Idea 3**: Train full LLM only English → train 25% top and 25% bottom → train 10% top and 10% bottom
- **gradually adapt with LoRA**: Insert lightweight LoRA adapters (rank 4, ∼0.5 % of full weights per module) at every third Transformer layer (e.g., layers 3, 6, 9, …).

**Comment by alex-shvets** [comment]
- English
- German, Spanish, French, Italian, Portuguese
- Polish, Turkish, Dutch, Czech, Hungarian


Model parameters from [open-sci experiments](https://docs.google.com/spreadsheets/d/1CgZyYLbINCajBaGz2sug4XA1ymLD3CZPiQ8BoeJ_L8s/edit?pli=1&gid=0#gid=0):

- C4 reference 1.3B cosine 300B 2M | optimal |   |   | Done | Low | 1.3B | 300B | cosine | 3,00E-03 | 3,00E-05 | 5000 | 1024 | 2048 | 143052 | 1,00000E+04 |   |   | 2,50008E+00 | 2,43192E+00 | 1,13807E+01 |   |   | Jenia | C4 | GPT-NeoX | JEDI | Reference hypers tuning | hypers matching 50B training | Jenia | 

Setup 1: 100-0 (English-rest)
Setup 2: 50-10-10-10-10-10 (English - German, Spanish, French, Italian, Portuguese)

300B tokens in total across all stages.

**Comment by inakiLakunza** [comment]
Hi, 
We have started our experiments running two different data configurations with a 2B version of the OpenSci model. Our idea is to introduce different languages through different stages of the training. The languages that @alex-shvets commented. For this, we have run the first stage of the training of these models, with 100B tokens each, the first one using all English data and the second one using 50B tokens for english and 10B for German, Spanish, French, Italian and Portuguese.

This sums up to ~47k training steps, and we have done evaluations for both models every 10k steps to check evolution. For now there is significant improvement of the first config against the second in English tasks, and vice versa for multilingual tasks, as it is expected. You can see the scores here: [results first stage](https://docs.google.com/spreadsheets/d/1ijmWxuJxbP96_s-26qRXTTnHt1MgzvkB3Lti0ylZZIM/edit?usp=sharing)

But we want to check how this situation evolves as the next epochs are done, since we will use the same amount of data for each language in both configurations. And so, this way we will be able to check what is the best approach for adding new languages to a pre-training (since the beginning or after there is quite a large english knowledge).

We are also thinking which set of tasks we should use to evaluate the models in multilingual setups, so we are open to any suggestions.

**Comment by spyysalo** [comment]
Progress currently blocked by Megatron apparently leaking CPU memory on MN5 (@inakiLakunza)

**Comment by alex-shvets** [comment]
- Setup1: English (100B) -> English (15B) + top5 EU -> English (5B) + top5 EU + next top5 EU (55B).
- Setup2: English (50B) + top5 EU (50B) -> English (35B) + top5 EU (35B) + next top5 EU (35B) -> English (35B) + top5 EU (25B) + next top5 EU (20B)
- Ad-hoc: Standard non-staged WSD training with the same amount of tokens (English - 120B, top5 EU - 110B, next top5 EU - 55B)

Observations:
- Starting training only with English data and adding other languages performs the poorest.
- Staged training, starting with English and the top 5 EU allows smoother integration of new languages.
- Last checkpoints for Setup2 and Ad-hoc perform similarly when they start from the same learning rate.
- Setup2 allows for earlier improvements in the newly introduced languages (than Ad-hoc in corresponding languages), with fast recoveries in previously added languages. However, in the long run, they converge to the same scores.

<img width="3600" height="3600" alt="Image" src="https://github.com/user-attachments/assets/51912abb-24c0-4684-a912-2d96fc921bf0" />

**Comment by alex-shvets** [comment]
<img width="3600" height="3600" alt="Image" src="https://github.com/user-attachments/assets/4b313600-4610-4b1e-8fe3-fa3760da92c8" />

**Comment by alex-shvets** [comment]
We continued with the 4th stage, adding 25 new languages (**36 in total**). We also implemented partial layer-freezing to preserve the multilingual core when switching the language distribution.

IMPORTANT: The normal training (red line below) was conducted using **only 11 languages**. So, it was included to show a trade-off when 11 or 36 languages are considered.

Evaluation with Belebele:

<img width="3600" height="3600" alt="Image" src="https://github.com/user-attachments/assets/924d7a97-d4f9-4819-9af5-b30ec0e70c24" />

Layer-freezing evaluated with Belebele. The freezing has been applied in layers 3-20:

<img width="3600" height="3600" alt="Image" src="https://github.com/user-attachments/assets/63b3eb81-7784-439c-a037-3cc11e25488c" />

Evaluation with Goldenswag (new languages are evaluated):

<img width="3600" height="3600" alt="Image" src="https://github.com/user-attachments/assets/aa39bd13-1885-4cfe-bd94-9eae6052273d" />

Setup 2 only, with and without freezing:

<img width="3600" height="2100" alt="Image" src="https://github.com/user-attachments/assets/a4ba1b00-cefc-428b-af45-d43e428daf64" />

**Comment by alex-shvets** [comment]
The 4th epoch was rather unsuccessful. Freezing the layers helped only partially. Conclusion: we shall not add too many languages at once.

[OELLM Results - Stage 4 42EN.pdf](https://github.com/user-attachments/files/23475715/OELLM.Results.-.Stage.4.42EN.pdf)

**Comment by alex-shvets** [comment]
The 5th stage only with English data harms other languages, and doesn't add much to English performance:

[OELLM Results - Stage 5 42EN.pdf](https://github.com/user-attachments/files/23475825/OELLM.Results.-.Stage.5.42EN.pdf)

**Comment by alex-shvets** [comment]
We are adding three more setups with only a few stages each, and with other languages.
We are going to train 2B models with 500B tokens.


Setup v2_1: Sequential addition with good coverage of English at all stages, but with a reduction of weight within the stage to favour multilinguality (90% -> 74% -> 50%):
```
    stage_weights["en"] = [0.434285714, 0.282857143, 0.282857143]
    stage_weights["math"] = [0.434285714, 0.282857143, 0.282857143]
    stage_weights["code"] = [0.434285714, 0.282857143, 0.282857143]
    stage_weights["zh"] = [0.140740741, 0.259259259, 0.6]
    stage_weights["es"] = [0.140740741, 0.259259259, 0.6]
    stage_weights["de"] = [0.140740741, 0.259259259, 0.6]
    stage_weights["fr"] = [0.140740741, 0.259259259, 0.6]
    stage_weights["ru"] = [0.140740741, 0.259259259, 0.6]
    stage_weights["it"] = [0, 0.1, 0.9]
    stage_weights["pl"] = [0, 0.1, 0.9]
    stage_weights["du"] = [0, 0.1, 0.9]
    stage_weights["tu"] = [0, 0.1, 0.9]
    stage_weights["hu"] = [0, 0.1, 0.9]
    stage_weights["instructions"] = [0.337777778, 0.266222222, 0.396]
```

Setup v2_2 - All languages from the beginning, with a reduction of English's weight within the stage to favour multilinguality (80% -> 60%):
```
    stage_weights["en"] = [0.575714286, 0.424285714]
    stage_weights["math"] = [0.575714286, 0.424285714]
    stage_weights["code"] = [0.575714286, 0.424285714]
    stage_weights["zh"] = [0.4, 0.6]
    stage_weights["es"] = [0.4, 0.6]
    stage_weights["de"] = [0.4, 0.6]
    stage_weights["fr"] = [0.4, 0.6]
    stage_weights["ru"] = [0.4, 0.6]
    stage_weights["it"] = [0.1, 0.9]
    stage_weights["pl"] = [0.1, 0.9]
    stage_weights["du"] = [0.1, 0.9]
    stage_weights["tu"] = [0.1, 0.9]
    stage_weights["hu"] = [0.1, 0.9]
    stage_weights["instructions"] = [0.505, 0.495]
```

Setup v2_3 - All languages from the beginning, with an enhancement of English's weight within the stage to favour core skills but retaining multilinguality (60% -> 80%):
```
    stage_weights["en"] = [0.371428571, 0.628571429]
    stage_weights["math"] = [0.371428571, 0.628571429]
    stage_weights["code"] = [0.371428571, 0.628571429]
    stage_weights["zh"] = [0.666666667, 0.333333333]
    stage_weights["es"] = [0.666666667, 0.333333333]
    stage_weights["de"] = [0.666666667, 0.333333333]
    stage_weights["fr"] = [0.666666667, 0.333333333]
    stage_weights["ru"] = [0.666666667, 0.333333333]
    stage_weights["it"] = [0.5, 0.5]
    stage_weights["pl"] = [0.5, 0.5]
    stage_weights["du"] = [0.5, 0.5]
    stage_weights["tu"] = [0.5, 0.5]
    stage_weights["hu"] = [0.5, 0.5]
    stage_weights["instructions"] = [0.45, 0.55]
```

**Comment by alex-shvets** [comment]
Evals EN:

<img width="2400" height="1200" alt="Image" src="https://github.com/user-attachments/assets/edb9bef8-27c7-4747-8221-7cdd9d62d981" />

<img width="2400" height="1200" alt="Image" src="https://github.com/user-attachments/assets/eea5e204-046c-4ed0-9c92-a83dbea0b89b" />

Evals top5 non-EN:

<img width="2400" height="1200" alt="Image" src="https://github.com/user-attachments/assets/d9bd7df3-1a6a-4112-9cc6-82cee8b45879" />

<img width="2400" height="1200" alt="Image" src="https://github.com/user-attachments/assets/3cd6e29a-c42f-48fa-bb1a-5f0c79930791" />

**Comment by alex-shvets** [comment]
We are currently training models for 49 languages + code, with the following three stages: EN 80-75-50, mid-resource 14-26-60, and low-resource 0-10-90.

- 7B and MoE with 500BT.
- MoE with 10TT.

**Comment by jsaizant** [comment]
In the 7B model training, after completing Stages 1 and 2, we started training Stage 3 and evaluated it up to Step 116K (global Step ~198K) using multilingual (MMLU, Belebele, Goldenswag, Xstorycloze) and monolingual (Spanish, English, Chinese) tasks. 
From Stage 2 onwards, we decreased the EN ratio from 75% to 50%, and increased the proportion of mid-resourced data from 26% to 60%, and low resourced data from 10% to 90%.
We observed that performance in low-resource languages increased (see Goldenswag) or remained stable (see MMLU or Belebele), while performance in mid-resourced and EN languages decreased, as would be expected when the training data changes during training.

<img width="2387" height="853" alt="Image" src="https://github.com/user-attachments/assets/88347483-fdf1-4c54-8348-58aa7843541d" />
<img width="2387" height="884" alt="Image" src="https://github.com/user-attachments/assets/8688ea09-2a29-4bf5-90dc-375d0bcae73a" />
<img width="2387" height="884" alt="Image" src="https://github.com/user-attachments/assets/e7e8342f-47ad-4c94-a2c2-205aeaefd468" />
<img width="2387" height="884" alt="Image" src="https://github.com/user-attachments/assets/4977567c-d87b-4ed1-b2df-3a2d7d9bb201" />
<img width="2387" height="884" alt="Image" src="https://github.com/user-attachments/assets/05ee46ad-0971-40c5-951f-f80e9a80da8a" />

**Comment by jsaizant** [comment]
Preliminar conclusions and insights from the experiments done on [Task #23](https://github.com/OpenEuroLLM/Taskboard/issues/23).
TLDR: The staged approach shows a clear advantage over the ad-hoc approach: although distribution changes (smooth or harsh) cause initial performance dips at each stage, the approach recovers and delivers gains overall. Encouragingly, these improvements hold across both the 2B and 7B dense models and the MoE models, suggesting the benefit is robust rather than a fluke of one architecture. Exhaustive evaluation across more languages is still pending on all checkpoints, so we don't yet know exactly how large the advantage is, but the signal so far is promising.

| Conclusion                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | Exact reason (Where/ which experiment)                                                                                                                                                                                                                                                                                                                                  |
| ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **1\. Start multilingual rather than English-Only<br>Beginning with a multilingual base, even a short one, is better than with an English-only base. It ensures smoother integration of new languages later on.**                                                                                                                                                                                                                                                                                                                                                                  | alex-shvets, Sep 2 [https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3244441742](https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3244441742)                                                                                                                                                                                             |
| **2\. Rapid integration with transfer learning<br>When new languages are introduced to a model that already possesses a small but sufficient multilingual base, the performance boost for those new languages is incredibly large and rapid.**                                                                                                                                                                                                                                                                                                                                     | alex-shvets, Sep 2 [https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3244441742](https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3244441742)                                                                                                                                                                                             |
| **3\. Data buffers mitigate catastrophic forgetting of a language.<br>Models suffer performance degradation in previously learned languages if those languages are completely omitted from new training phases. We tried by maintaining a small 0.5% token buffer per language in untargeted languages, not to harm the target language. However, this small buffer per language didn't help as much as gradually increasing the ratio of the language in each epoch, while decreasing English smoothly.**                                                                                                         | Internal experiments.                                                                                                                                                                                                                                                                                                                                                   |
| **4\. Distribution changes cause temporary drops.<br>Drastic changes in data distribution (such as aggressively shifting language ratios or adding massive amounts of new languages) cause an performance drop. After warmup, lost performance is regained while new languages gain performance.**                                                                                                                                                                                                                                                                                 | alex-shvets, Dec 9 [https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3631360730](https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3631360730)                                                                                                                                                                                             |
| **5\. Layer freezing offers short-term stability but disrupts performance long-term.<br>Freezing the middle layers during the introduction of new languages successfully prevents sudden performance drops in existing languages. However, it slows down the overall learning rate. In the long run, the final results are highly similar to not freezing at all (where the model drops sharply but recovers). Furthermore, freezing during warmup and unfreezing later fails, as the sudden learning rate shift combined with a new data distribution destabilizes the network.** | alex-shvets, Oct 14, 2025 [https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3400709654](https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3400709654) ; alex-shvets, Nov 11, 2025 [https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3516029856](https://github.com/OpenEuroLLM/Taskboard/issues/23#issuecomment-3516029856) |                                                                                                                                                                                                                                                                                                                                                      |
